scripts/release/relnotes.py

```diff
@@ -0,0 +1,139 @@
+  # sort. Note that we have a version "0.2.2b" so we just convert any non-digit
+  # segment into 9999.
+  return [int(s) if s.isdigit() else 9999 for s in v.split(".")]
+
+
+def get_latest_release():
+  """Discovers the latest stable release name from GCS."""
+  # The following call lists all version bases (the X.X.X part), including
+  # RC-only and rolling-only versions.
+  versions = [prefix.rstrip("/") for json in _paginated_get(GCS_BASE_URL)
+              if "prefixes" in json
+              for prefix in json["prefixes"]]
+  # Sort the versions from latest to earliest.
+  versions.sort(key=_version_sort_key, reverse=True)
+  for version in versions:
+    # The following call identifies the first version that has a stable release.
+    if any("items" in json for json in
+           _paginated_get(f"{GCS_BASE_URL}&prefix={version}/release/")):
+      return version
+  raise ValueError("couldn't find a single release version!")
+
+
+def git(*args):
+  """Runs git as a subprocess, and returns its stdout as a list of lines."""
+  return subprocess.check_output(["git"] + list(args)).decode(
+    "utf-8").strip().split("\n")
+
+
+def extract_relnotes(commit_message_lines):
+  """Extracts relnotes from a commit message (passed in as a list of lines)."""
+  relnote_lines = []
+  in_relnote = False
+  for line in commit_message_lines:
+    if line == "" or line.startswith("PiperOrigin-RevId:"):
+      in_relnote = False
+    m = re.match(r"^RELNOTES(?:\[(INC|NEW)\])?:", line)
+    if m is not None:
+      in_relnote = True
+      line = line[len(m[0]):]
+      if m[1] == "INC":
+        line = "**[Incompatible]** " + line.strip()
+    line = line.strip()
+    if in_relnote and line != "":
+      relnote_lines.append(line)
+  relnote = " ".join(relnote_lines)
+  relnote_lower = relnote.strip().lower().rstrip(".")
+  if relnote_lower == "n/a" or relnote_lower == "none":
+    return None
+  return relnote
+
+
+def get_relnotes_between(base, head):
+  """Gets all relnotes for commits between `base` and `head`."""
+  commits = git("rev-list", f"{base}..{head}", "--grep=RELNOTES")
+  relnotes = []
+  rolled_back_commits = set()
+  # We go in reverse-chronological order, so that we can identify rollback
+  # commits and ignore the rolled-back commits.
+  for commit in commits:
+    if commit in rolled_back_commits:
+      continue
+    lines = git("show", "-s", commit, "--pretty=format:%B")
+    m = re.match(r"^Automated rollback of commit ([\dA-Fa-f]+)", lines[0])
+    if m is not None:
+      rolled_back_commits.add(m[1])
+      # The rollback commit itself is also skipped.
+      continue
+    relnote = extract_relnotes(lines)
+    if relnote is not None:
+      relnotes.append(relnote)
+  return relnotes
+
+
+if __name__ == "__main__":
+  # Get the last stable release.
+  last_release = get_latest_release()
+  print("last_release is", last_release)
+  last_release_branch = "origin/release-" + last_release
```

Comment:
This assume `origin/release-<version>` is already fetched, but might not be true.
Also maybe it's slightly better to calculate the base commit from the tag instead a the branch?
We can do
```
git fetch origin 5.3.2
git merge-base FETCH_HEAD HEAD
```

Reply:
changed to fetch the tag first